include/beman/execution26/detail/atomic_intrusive_stack.hpp

```diff
@@ -0,0 +1,83 @@
+// include/beman/execution26/detail/atomic_intrusive_stack.hpp        -*-C++-*-
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef INCLUDED_BEMAN_EXECUTION26_DETAIL_ATOMIC_INTRUSIVE_STACK
+#define INCLUDED_BEMAN_EXECUTION26_DETAIL_ATOMIC_INTRUSIVE_STACK
+
+#include <beman/execution26/detail/intrusive_queue.hpp>
+
+#include <atomic>
+#include <cassert>
+#include <optional>
+
+namespace beman::execution26::detail {
+
+template <auto Next>
+class atomic_intrusive_stack;
+
+//! @brief  This data structure is an intrusive stack that can be used in a lock-free manner.
+//!
+//! The stack is implemented as a singly linked list where the head is an atomic pointer to the first item in
+//! the stack.
+//! try_push() is a lock-free operation that tries to push an item to the stack. If the stack is empty, it
+//! returns nullptr and the item is pushed to the stack.
+//! This stack has a closed state, which is indicated by the head pointing to the stack itself. In this state,
+//! try_push() returns std::nullopt and the stack is not modified.
+//! pop_all_and_shutdown() is a lock-free operation that pops all items from the stack and returns them in a
+//! queue. If the stack is empty, it returns an empty queue.
+//!
+//! We use this stack in the split implementation to store the listeners that are waiting for the operation to
+//! complete.
+//!
+//! @tparam Item  The type of the item in the stack.
+//! @tparam Next  The pointer to the next item in the stack.
+template <class Item, Item* Item::*Next>
```

Comment:
In the past compilers were pretty bad optimising anything around pointer to members. They tend to be way better at optimising around inline functions and the parameter could be something like

    decltype([](auto* node){ return node->next; })
  
This lambda could also be a default argument if the objective is avoiding the need to specify it. Admittedly, using this approach is more involved when actually using the data structure. This comment is primarily for consideration and doesn't necessarily need to be addressed.

Reply:
I will investigate code gen. Now I am curious